project/sbt-changeling/Changeling.scala

```diff
@@ -0,0 +1,280 @@
+   *   The base directory
+   * @param children
+   *   The names of children directories that should exist in `base`
+   */
+  def ensureDirStructureExists(base: File, children: Seq[String]): File = {
+    val childOfBase: String => File = base / _
+    val leafDirs = children.map(normalizeFileName.andThen(childOfBase))
+    IO.createDirectories(leafDirs)
+    base
+  }
+
+  /**
+   * Render the directory unreleased directory structure as a markdown file
+   *
+   * @param changeKinds
+   *   The supported kinds of cnhages, establishes order of change sections
+   * @param version
+   *   The version to be released, used as heading of realease notes
+   * @param unreleasedDir
+   *   The directory from which to read the unreleased change entries
+   * @param relaseNotes
+   *   The file into which the release notes will be written
+   * @returns
+   *   The TrelaseNotes
+   */
+  def renderReleaseNotes(
+      changeKinds: Seq[String],
+      version: String,
+      unreleasedDir: File,
+      releaseNotes: File): File = {
+
+    val changeDirOrder: File => Int = { d =>
+      changeKinds.indexOf(deNormalizeFileName(d.base.toString))
+    }
+
+    // The heading for the release notes
+    val versionHeading = s"## ${version} - ${LocalDate.now}"
+    // To render the directory structure described above into the required flatfile
+    // we just have to flatten the directory tree with the following transformations
+    //
+    // - transforming the change directory name into an H3 heading
+    // - get the contnet of each file in the change directory, and make it a
+    //   bullet list item
+    val changeSections = IO
+      .listFiles(unreleasedDir)
+      .sortBy(changeDirOrder)
+      .flatMap { changeDir =>
+        IO.listFiles(changeDir) match {
+          case Array() => Seq() // Omit sections with no entries
+          case changeEntries => {
+            val heading = s"### ${deNormalizeFileName(changeDir.base.toString)}"
+            val changes = changeEntries.map(readFile.andThen(mdBulletItem))
+            // We add an empty line between the heading and the changes and after the changes
+            (heading +: "" +: changes) ++ Seq("")
+          }
+        }
+      }
+    val notes = (versionHeading +: "" +: changeSections).mkString("\n")
+    IO.write(releaseNotes, notes)
+    releaseNotes
+  }
+
+  /**
+   * Prefix the content of `newFile` to the content of `target` and write the result to `target`
+   */
+  def concatFiles(front: File, target: File): Unit = {
+    val content = s"${IO.read(front)}\n${IO.read(target)}"
+    IO.write(target, content)
+  }
+
+  /**
+   * Move all the contents of the cange directories in `unrelasedDir` into a "trash" dir in `resourceDir`
+   *
+   * @return
+   *   the trash directory crated
+   */
+  def cleanChanges(resourceDir: File, unreleasedDir: File): Unit = {
+    // So we can recover dirs if needed
+    val trashDir = resourceDir / "changeling-trash" / LocalDate.now.toString
```

Comment:
I'm not clear yet how this would work when cutting a release. Are we committing the trashed files back into `main`?

Also, if we keep the trash, shouldn't we keep it by version, not by date?

Reply:
This is just a safeguard for when people run it on their local machines, just in case they, for instance, configure the .unreleased due to point to something important. It’s not meant to be kept or inspected at all, and plays no role in releases.

Reply:
sg, this was my only blocking concern